pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1225,6 +1163,83 @@ func (l *LibvirtDomainManager) findOrCreateVirDomain(
 	return virDomain, err
 }
 
+func (l *LibvirtDomainManager) handleDisksAttachment(
+	apiDomain *api.Domain,
+	virDomain cli.VirDomain,
+	vmi *v1.VirtualMachineInstance,
+) (*api.DomainSpec, error) {
+	logger := log.Log.Object(vmi)
+
+	oldSpec, err := getDomainSpec(virDomain)
+	if err != nil {
+		logger.Reason(err).Error("Parsing domain XML failed.")
+		return nil, err
+	}
+
+	// Look up all the disks to detach
+	for _, detachDisk := range getDetachedDisks(oldSpec.Devices.Disks, apiDomain.Spec.Devices.Disks) {
+		logger.V(1).Infof("Detaching disk %s, target %s", detachDisk.Alias.GetName(), detachDisk.Target.Device)
+		detachBytes, err := xml.Marshal(detachDisk)
+		if err != nil {
+			logger.Reason(err).Error("marshalling detached disk failed")
+			return nil, err
+		}
+		err = virDomain.DetachDeviceFlags(strings.ToLower(string(detachBytes)), affectDeviceLiveAndConfigLibvirtFlags)
+		if err != nil {
+			logger.Reason(err).Error("detaching device")
+			return nil, err
+		}
+	}
+	// Look up all the disks to attach
+	for _, attachDisk := range getAttachedDisks(oldSpec.Devices.Disks, apiDomain.Spec.Devices.Disks) {
+		allowAttach, err := checkIfDiskReadyToUse(getSourceFile(attachDisk))
+		if err != nil {
+			return nil, err
+		}
+		if !allowAttach {
+			continue
+		}
+		logger.V(1).Infof("Attaching disk %s, target %s", attachDisk.Alias.GetName(), attachDisk.Target.Device)
+		// set drivers cache mode
+		err = converter.SetDriverCacheMode(&attachDisk, l.directIOChecker)
+		if err != nil {
+			return nil, err
+		}
+		err = converter.SetOptimalIOMode(&attachDisk)
+		if err != nil {
+			return nil, err
+		}
+
+		attachBytes, err := xml.Marshal(attachDisk)
+		if err != nil {
+			logger.Reason(err).Error("marshalling attached disk failed")
+			return nil, err
+		}
+		err = virDomain.AttachDeviceFlags(strings.ToLower(string(attachBytes)), affectDeviceLiveAndConfigLibvirtFlags)
+		if err != nil {
+			logger.Reason(err).Error("attaching device")
+			return nil, err
+		}
+	}
+
+	// Resize and notify the VM about changed disks
+	for _, disk := range apiDomain.Spec.Devices.Disks {
+		if shouldExpandOnline(virDomain, disk) {
+			possibleGuestSize, ok := possibleGuestSize(disk)
+			if !ok {
+				logger.Reason(err).Warningf("Failed to get possible guest size from disk %v", disk)
+				break
+			}
+			err := virDomain.BlockResize(getSourceFile(disk), uint64(possibleGuestSize), libvirt.DOMAIN_BLOCK_RESIZE_BYTES)
+			if err != nil {
+				logger.Reason(err).Errorf("libvirt failed to expand disk image %v", disk)
+			}
+		}
+	}
+
+	return oldSpec, nil
```

Comment:
The old spec better be passed in to the method, not returned here. It makes little sense that the method that handles the drives is returning the old spec.

Reply:
Indeed, I'll change it.